flit_core/flit_core/common.py

```diff
@@ -127,26 +138,26 @@ def get_docstring_and_version_via_ast(target):
     Return a tuple like (docstring, version) for the given module,
     extracted by parsing its AST.
     """
-    # read as bytes to enable custom encodings
-    with target.file.open('rb') as f:
-        node = ast.parse(f.read())
-    for child in node.body:
-        # Only use the version from the given module if it's a simple
-        # string assignment to __version__
-        is_version_str = (
-                isinstance(child, ast.Assign)
-                and any(
-                    isinstance(target, ast.Name)
-                    and target.id == "__version__"
-                    for target in child.targets
-                )
-                and isinstance(child.value, ast.Str)
-        )
-        if is_version_str:
-            version = child.value.s
-            break
-    else:
-        version = None
+    version = None
+    for target_path in target.version_files:
+        # read as bytes to enable custom encodings
+        with target_path.open('rb') as f:
+            node = ast.parse(f.read())
+        for child in node.body:
+            # Only use the version from the given module if it's a simple
+            # string assignment to __version__
+            is_version_str = (
+                    isinstance(child, ast.Assign)
+                    and any(
+                        isinstance(target, ast.Name)
+                        and target.id == "__version__"
+                        for target in child.targets
+                    )
+                    and isinstance(child.value, ast.Str)
+            )
+            if is_version_str:
+                version = child.value.s
+                break
     return ast.get_docstring(node), version
```

Comment:
I think the docstring should always come from the main file (i.e. `__init__.py` if this is a package), rather than the version file. The first line of this docstring is used as the short description for PyPI by default.

Reply:
@takluyver I believe that's how this works as-implemented. The order of lookup is intentionally crafted with the `__init__.py` as the last entry `target.version_files`, and the loop doesn't short circuit. The docstring lookup only happens after the loop is completed, when `node` will always have the contents of `__init__.py`.

The existing test case [1][] passes, and contains the docstring from the fixture's `__init__.py`. I went ahead and pushed an additional change to add an "imposter" docstring to the `_version.py` fixture, to help show that it's not being used/selected.

[1]: https://github.com/pypa/flit/blob/fc640d2c79441bc8a37e5098db898ecd42bcc44f/flit_core/flit_core/tests/test_common.py#L83

Reply:
You're right, thanks. I had scanned the code and assumed the `break` would stop at the first file containing `__version__`, but of course it only breaks out of the inner loop.